src/coprocessor/mod.rs

```diff
@@ -281,38 +281,75 @@ pub(crate) mod test {
             Ok((c_ptr, input_env.clone(), input_cont.clone()))
         }
 
-        fn synthesize_lem_simple<CS: ConstraintSystem<F>>(
+        fn synthesize_lem<CS: ConstraintSystem<F>>(
             &self,
             cs: &mut CS,
             g: &GlobalAllocator<F>,
-            _s: &LEMStore<F>,
-            not_dummy: &Boolean,
+            s: &LEMStore<F>,
+            _not_dummy: &Boolean,
             args: &[AllocatedPtr<F>],
-        ) -> Result<AllocatedPtr<F>, SynthesisError> {
-            use crate::lem::Tag::Expr;
+            env: &AllocatedPtr<F>,
+            cont: &AllocatedPtr<F>,
+        ) -> Result<Vec<AllocatedPtr<F>>, SynthesisError> {
             let a = &args[0];
             let b = &args[1];
 
             let allocated_num_tag = g
-                .get_allocated_const(Expr(ExprTag::Num).to_field())
+                .get_allocated_const(LEMTag::Expr(ExprTag::Num).to_field())
                 .expect("Num tag should have been allocated");
-            implies_equal(
-                &mut cs.namespace(|| "first arg tag"),
-                not_dummy,
+            let a_tag_is_num = alloc_equal(
+                &mut cs.namespace(|| "fst tag is num"),
                 a.tag(),
                 allocated_num_tag,
-            );
-            implies_equal(
-                &mut cs.namespace(|| "second arg tag"),
-                not_dummy,
+            )?;
+            let b_tag_is_num = alloc_equal(
+                &mut cs.namespace(|| "snd tag is num"),
                 b.tag(),
                 allocated_num_tag,
-            );
+            )?;
+            let tags_are_num = and(
+                &mut cs.namespace(|| "tags are num"),
+                &a_tag_is_num,
+                &b_tag_is_num,
+            )?;
 
             // a^2 + b = c
             let a2 = mul(&mut cs.namespace(|| "square"), a.hash(), a.hash())?;
             let c = add(&mut cs.namespace(|| "add"), &a2, b.hash())?;
-            AllocatedPtr::alloc_tag(cs, ExprTag::Num.to_field(), c)
+            let expr_ok = AllocatedPtr::from_parts(allocated_num_tag.clone(), c);
+
+            let allocated_zero = g
+                .get_allocated_const(F::ZERO)
+                .expect("zero should have been allocated");
+            let expr_err =
+                AllocatedPtr::from_parts(allocated_num_tag.clone(), allocated_zero.clone());
+            let expr = AllocatedPtr::pick(
+                &mut cs.namespace(|| "pick expr"),
+                &tags_are_num,
+                &expr_ok,
+                &expr_err,
+            )?;
+
+            // the following is a temporary shim for compatibility with Lurk Alpha
+            // otherwise we could just have a pointer whose value is zero
```

Comment:
It is, but I think the best way to make this easy to maintain is the following:
- isolate the expected error behavior demanded by Lurk Alpha in a unit test,
- mark that compatibility requirement here by making an allusion to Lurk Alpha (as you already do) and pointing to this test for details.

Reply:
LEM infra won't allow this to go on with the wrong values: https://github.com/lurk-lab/lurk-rs/blob/master/src/lem/circuit.rs#L632-L636 and I prefer to keep it that way.

But I'm already committed to changing those (see #716)

For reference, these were introduced in #674

Reply:
Just to clarify: my comment wasn't hinting at any sort of desired change in behavior. It was suggesting a way to present information about that behavior, as it is today.

Reply:
I see. I can do it on another PR

Reply:
Never mind then.